src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.LazySyntaxTree.cs

```diff
@@ -0,0 +1,188 @@
+                    root = Interlocked.CompareExchange(ref _lazyRoot, root, null) ?? root;
+                }
+
+                return root;
+            }
+
+            public override bool TryGetRoot([NotNullWhen(true)] out CSharpSyntaxNode? root)
+            {
+                root = _lazyRoot;
+                return root != null;
+            }
+
+            public override bool HasCompilationUnitRoot
+            {
+                get
+                {
+                    return true;
+                }
+            }
+
+            public override CSharpParseOptions Options
+            {
+                get
+                {
+                    return _options;
+                }
+            }
+
+            [Obsolete("Obsolete due to performance problems, use CompilationOptions.SyntaxTreeOptionsProvider instead", error: false)]
+            public override ImmutableDictionary<string, ReportDiagnostic> DiagnosticOptions => _diagnosticOptions;
+
+            public override SyntaxReference GetReference(SyntaxNode node)
+            {
+                return new SimpleSyntaxReference(node);
+            }
+
+            public override SyntaxTree WithRootAndOptions(SyntaxNode root, ParseOptions options)
+            {
+                if (ReferenceEquals(_lazyRoot, root) && ReferenceEquals(_options, options))
+                {
+                    return this;
+                }
+
+                return new ParsedSyntaxTree(
+                    textOpt: null,
+                    _text.Encoding,
+                    _text.ChecksumAlgorithm,
+                    _path,
+                    (CSharpParseOptions)options,
+                    (CSharpSyntaxNode)root,
+                    _directives,
+                    _diagnosticOptions,
+                    cloneRoot: true);
+            }
+
+            public override SyntaxTree WithFilePath(string path)
+            {
+                if (_path == path)
+                {
+                    return this;
+                }
+
+                if (TryGetRoot(out var root))
+                {
+                    return new ParsedSyntaxTree(
+                        _text,
+                        _text.Encoding,
+                        _text.ChecksumAlgorithm,
+                        path,
+                        _options,
+                        root,
+                        GetDirectives(),
+                        _diagnosticOptions,
+                        cloneRoot: true);
+                }
+                else
+                {
+                    return new LazySyntaxTree(_text, path, _options, _diagnosticOptions);
+                }
+            }
+
+            [Obsolete("Obsolete due to performance problems, use CompilationOptions.SyntaxTreeOptionsProvider instead", error: false)]
+            public override SyntaxTree WithDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic> options)
```

Comment:
Having this be `options` instead of `diagnsticOptions` makes the impl a bit harder to read here. Particularly when we're passing `_options` and `options` to a ctor. Looks like a bug until you read deeper into the code.

Reply:
This one was also taken from existing code:
https://github.com/dotnet/roslyn/blob/afd10305a37c0ffb2cfb2c2d8446154c68cfa87a/src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.ParsedSyntaxTree.cs#L162-L163